internal/fwserver/server.go

```diff
@@ -9,4 +14,41 @@ import (
 // response type conversions.
 type Server struct {
 	Provider tfsdk.Provider
+
+	// providerSchema is the cached Provider Schema for RPCs that need to
+	// convert configuration data from the protocol. If not found, it will be
+	// fetched from the Provider.GetSchema() method.
+	providerSchema *tfsdk.Schema
+
+	// providerSchemaDiags is the cached Diagnostics obtained while populating
+	// providerSchema. This is to ensure any warnings or errors are also
+	// returned appropriately when fetching providerSchema.
+	providerSchemaDiags diag.Diagnostics
+
+	// providerSchemaMutex is a mutex to protect concurrent providerSchema
+	// access from race conditions.
+	providerSchemaMutex sync.Mutex
+}
+
+// ProviderSchema returns the Schema associated with the Provider. The Schema
+// and Diagnostics are cached on first use.
+func (s *Server) ProviderSchema(ctx context.Context) (*tfsdk.Schema, diag.Diagnostics) {
+	logging.FrameworkTrace(ctx, "Checking ProviderSchema lock")
+	s.providerSchemaMutex.Lock()
+
+	if s.providerSchema != nil {
+		s.providerSchemaMutex.Unlock()
+		return s.providerSchema, nil
```

Comment:
Should this return also `s.ProviderSchemaDiags`?

Unless the intention is to report potential diag-reported warnings/errors, only the first time?

Reply:
Yes, it should, thanks for noticing!